### PR TITLE
Enable C++ branch coverage when using gcov 8 or later.

### PR DESCRIPTION
**Pull request by c-mita**
Branch coverage was disabled due to a gcov bug, but this was fixed for
GCC 8: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84879.

It doesn't make sense to disable all C++ branch coverage for everyone
because of a bug in an old GCC version.

### CODE REVIEW

**Comment by c-mita** [comment]
There's quite a bit of duplication of helper functions in the various coverage related shell tests; I'll have a go at reducing that in a follow up change.

File: src/test/shell/bazel/bazel_cc_code_coverage_test.sh [hunk 1]
```diff
@@ -376,6 +376,11 @@ function test_cc_test_coverage_gcov() {
       output_file_json="output_file.json"
       zcat $agcda $tgcda $dagcda > $output_file_json
 
+      # Remove all branch information from the json.
+      # We disable branch coverage for gcov 7 and it is easier to remove it
+      # than it is to put it back.
+      # Replace "branch": [..] with "branch": [] - (non-empty [..])
```
Comment (meisterT) [inline src/test/shell/bazel/bazel_cc_code_coverage_test.sh:382]: nit: I assume you meant to write `branches` here?

File: src/test/shell/bazel/bazel_coverage_cc_test_gcc.sh [hunk 2]
```diff
@@ -110,6 +110,24 @@ was not found in actual coverage report:
...
+# file, accounting for the fact that branch coverage is disabled for gcov v7
+#
+# - expected coverage The expected result that must be included in the output.
+# - output_file       The location of the coverage output file.
+function assert_cc_coverage_result() {
+  local expected_coverage="${1}"; shift
+  local output_file="${1}"; shift
+
+  # we disable branch coverage when using gcov 7 so we should strip all branch
+  # information from the "expected" result before checking.
```
Comment (meisterT) [inline src/test/shell/bazel/bazel_coverage_cc_test_gcc.sh:123]: nit: please include an example `gcov -v` output in the comment here and explain that you are extracting the major version

Comment (meisterT) [reply src/test/shell/bazel/bazel_coverage_cc_test_gcc.sh:123]: same in all places below

File: tools/test/collect_cc_coverage.sh [hunk 3]
```diff
@@ -132,6 +132,12 @@ function gcov_coverage() {
               mkdir -p "${COVERAGE_DIR}/$(dirname ${gcno_path})"
               cp "$ROOT/${gcno_path}" "${COVERAGE_DIR}/${gcno_path}"
           fi
+
+          # Extract gcov's version: the output of `gcov --version` contains the
+          # version as a set of major-minor-patch numbers, of which we extract
+          # the major version.
+          gcov_major_version=$("${GCOV}" --version | sed -n -E -e 's/^.*\s([0-9]+)\.[0-9]+\.[0-9]+\s?.*$/\1/p')
```
Comment (meisterT) [inline tools/test/collect_cc_coverage.sh:139]: Why are you using a different mechanism to extract the major version here?

Comment (c-mita) [reply tools/test/collect_cc_coverage.sh:139]: It's simply replacing what was removed in https://github.com/bazelbuild/bazel/pull/18878.

This approach handles using `llvm-cov gcov`  - the tests don't need to worry about that.

Comment (c-mita) [reply tools/test/collect_cc_coverage.sh:139]: That said, when I come to consolidate things, I'll probably use the same approach everywhere (no reason not to).

But right now, I don't want to change things everywhere in the tests...

File: src/test/shell/bazel/bazel_cc_code_coverage_test.sh (hunk 1 above)
Comment (c-mita) [reply src/test/shell/bazel/bazel_cc_code_coverage_test.sh]: Fixed

File: src/test/shell/bazel/bazel_coverage_cc_test_gcc.sh (hunk 2 above)
Comment (c-mita) [reply src/test/shell/bazel/bazel_coverage_cc_test_gcc.sh:123]: Done.

**Review by meisterT — APPROVED** [review]

**Pull request closed without merging**